docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,201 @@
+- [Unhandled cases](#unhandled-cases)
+- [Implications to existing clients](#implications-to-existing-clients)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Overview
+
+Users of the server-side garbage collection need to determine if the garbage collection is done. For example:
+* Currently `kubectl delete rc` blocks until all the pods are terminating. To convert to use server-side garbage collection, kubectl has to be able to determine if the garbage collection is done.
+* [#19701](https://github.com/kubernetes/kubernetes/issues/19701#issuecomment-236997077) is a use case where the user needs to wait for all service dependencies garbage collected and their names released, before she recreates the dependencies.
+
+We define the garbage collection as "done" when all the dependents are deleted from the key-value store, rather than merely in the terminating state. There are two reasons: *i)* for `Pod`s, the most usual garbage, only when they are deleted from the key-value store, we know kubelet has released resources they occupy; *ii)* some users need to recreate objects with the same names, they need to wait for the old objects to be deleted from the key-value store. (This limitation is because we index objects by their names in the key-value store today.)
+
+Synchronous Garbage Collection is a best-effort (see [unhandled cases](#unhandled-cases)) mechanism that allows user to determine if the garbage collection is done: after the API server receives a deletion request of an owning object, the object keeps existing in the key-value store until all its dependents are deleted from the key-value store by the garbage collector.
+
+Tracking issue: https://github.com/kubernetes/kubernetes/issues/29891
+
+# API Design
+
+## Standard Finalizers
+
+We will introduce a new standard finalizer:
+
+```go
+const GCFinalizer string = “DeletingDependents”
+```
+
+This finalizer indicates the object is terminating and is waiting for its dependents whose `OwnerReference.BlockOwnerDeletion` is true get deleted.
+
+## OwnerReference
+
+```go
+OwnerReference {
+     ...
+     // If true, AND if the owner has the "DeletingDependents" finalizer, then the owner cannot be deleted from the key-value store until this reference is removed.
+     // Defaults to false.
+     // To set this field, a user needs "update" and "delete" permission of the owner, otherwise 422 (Unprocessable Entity) will be returned.
+     BlockOwnerDeletion *bool
+}
+```
+
+The initial draft of the proposal did not include this field and it had a security loophole: a user who is only authorized to update one resource can set ownerReference to block the synchronous GC of other resources. Requiring users to explicitly set `BlockOwnerDeletion` allows the master to properly authorize the request.
+
+## DeleteOptions
+
+```go
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to DefaultPropagationPolicy
+  DeletionPropagationPolicy *DeletionPropagationPolicy
```

Comment:
The field name can probably just be `PropagationPolicy`, it's redundant to say "delete" since it's in the DeleteOptions already. (Type name should still include "Delete")